stackable-webhook/src/lib.rs

```diff
@@ -0,0 +1,165 @@
+//! Utility types and functions to easily create ready-to-use webhook servers
+//! which can handle different tasks, for example CRD conversions. All webhook
+//! servers use HTTPS by default and provides options to enable HTTP to HTTPS
+//! redirection as well. This library is fully compatible with the [`tracing`]
+//! crate and emits multiple levels of tracing data.
+//!
+//! Most users will only use the top-level exported generic [`WebhookServer`]
+//! which enables complete control over the [Router] which handles registering
+//! routes and their handler functions.
+//!
+//! ```
+//! use stackable_webhook::{WebhookServer, Options};
+//! use axum::Router;
+//!
+//! let router = Router::new();
+//! let server = WebhookServer::new(router, Options::default());
+//! ```
+//!
+//! For some usages, complete end-to-end [`WebhookServer`] implementations
+//! exist. One such implementation is the [`ConversionWebhookServer`]. The
+//! only required parameters are a conversion handler function and [`Options`].
+//!
+//! This library additionally also exposes lower-level structs and functions to
+//! enable complete controll over these details if needed.
+use axum::Router;
+use snafu::{ResultExt, Snafu};
+use tracing::{debug, instrument};
+
+use crate::{options::RedirectOption, redirect::Redirector, tls::TlsServer};
+
+pub mod constants;
+pub mod options;
+pub mod redirect;
+pub mod servers;
+pub mod tls;
+
+// Selected re-exports
+pub use crate::{options::Options, servers::ConversionWebhookServer};
+
+/// A result type alias with the library-level [`Error`] type as teh default
+/// error type.
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+/// A generic webhook handler receiving a request and sending back a response.
+///
+/// This trait is not intended to be implemented by external crates and this
+/// library provides various ready-to-use implementations for it. One such an
+/// /// implementation is part of the [`ConversionWebhookServer`].
```

Comment:
Duplicate doc-comment thing

```suggestion
/// implementation is part of the [`ConversionWebhookServer`].
```